File: mask.go
package dotbot

type Mask uint64

const NumDots = BoardSize * BoardSize
const AllDots = (Mask(1) << NumDots) - 1

func InBounds(row, col int) bool {
	return 0 <= row && row < BoardSize && 0 <= col && col < BoardSize
}

func index(row, col int) uint {
	if !InBounds(row, col) {
		panic("Index out of bounds")
	}
	return uint(BoardSize)*uint(col) + uint(row)
}

func unIndex(index uint) (row, col int) {
	return int(index % uint(BoardSize)), int(index / uint(BoardSize))
}

func (mask Mask) Matches(pattern Mask) bool {
	return (mask & pattern) == pattern
}

func DotMask(row, col int) Mask {
	return Mask(1) << index(row, col)
}

func (mask Mask) Contains(row, col int) bool {
	return InBounds(row, col) && mask.Matches(DotMask(row, col))
}

func (mask *Mask) Add(row, col int) {
	*mask |= DotMask(row, col)
}

func (mask *Mask) Remove(row, col int) {
	*mask &= ^DotMask(row, col)
}

func (mask Mask) Count() (count int) {
	for mask != 0 {
		mask ^= (mask & -mask)
		count++
	}
	return
}

func (mask Mask) CountNeighbors(row, col int) int {
	count := 0
	if mask.Contains(row-1, col) {
		count++
	}
	if mask.Contains(row+1, col) {
		count++
	}
	if mask.Contains(row, col-1) {
		count++
	}
	if mask.Contains(row, col+1) {
		count++
	}
	return count
}

func (mask Mask) Partition(partitions chan Mask) {
	for row := 0; row < BoardSize; row++ {
		for col := 0; col < BoardSize; col++ {
			if mask.Contains(row, col) {
				partitions <- mask.buildPartition(row, col)
			}
		}
	}
	close(partitions)
}

func (mask *Mask) buildPartition(row, col int) (partition Mask) {
	stack := new(stack)

	visit := func(row, col int) {
		if mask.Contains(row, col) {
			mask.Remove(row, col)
			stack.push(row, col)
		}
	}

	visit(row, col)

	for stack.size > 0 {
		row, col = stack.pop()
		partition.Add(row, col)

		visit(row-1, col)
		visit(row+1, col)
		visit(row, col-1)
		visit(row, col+1)
	}
	return
}

func (mask Mask) String() string {
	board := new(Board)
	for row := 0; row < BoardSize; row++ {
		for col := 0; col < BoardSize; col++ {
			if mask.Contains(row, col) {
				board.SetColor(row, col, NotEmpty)
			}
		}
	}
	return board.String()
}

// Internal data structure for buildPartition().

type stack struct {
	data [NumDots]uint
	size int
}

func (s *stack) push(row, col int) {
	s.data[s.size] = index(row, col)
	s.size++
}

func (s *stack) pop() (row, col int) {
	s.size--
	r, c := unIndex(s.data[s.size])
	return int(r), int(c)
}

File: solver/src/mask.c
#include <stdio.h>
#include <stdlib.h>

#include "mask.h"

int Count(Mask mask) {
    int count = 0;
    while (mask) {
        mask ^= (mask & -mask);
        count++;
    }
    return count;
}

int CountNeighbors(Mask mask, int row, int col) {
	int count = 0;
	if (CONTAINS(mask, row-1, col)) {
		count++;
	}
	if (CONTAINS(mask, row+1, col)) {
		count++;
	}
	if (CONTAINS(mask, row, col-1)) {
		count++;
	}
	if (CONTAINS(mask, row, col+1)) {
		count++;
	}
    return count;
}

void Partition(Mask mask, Queue *q) {
    int row, col;
    for (row = 0; row < BoardSize; row++) {
        for (col = 0; col < BoardSize; col++) {
            if (CONTAINS(mask, row, col)) {
                Push(q, (void*)buildPartition(&mask, 0, row, col));
            }
        }
    }
}

Mask buildPartition(Mask *mask, Mask p, int row, int col) {
    *mask = REMOVE(*mask, row, col);
    p = ADD(p, row, col);

    if (CONTAINS(*mask, row-1, col)) {
        p = buildPartition(mask, p, row-1, col);
    }
    if (CONTAINS(*mask, row+1, col)) {
        p = buildPartition(mask, p, row+1, col);
    }
    if (CONTAINS(*mask, row, col-1)) {
        p = buildPartition(mask, p, row, col-1);
    }
    if (CONTAINS(*mask, row, col+1)) {
        p = buildPartition(mask, p, row, col+1);
    }

	return p;
}

void DFS(Mask mask, Queue *paths) {
    // seen[startPoint][endPoints]
    // This uniquely identifies a path through a non cyclic mask.
    int seen[NumDots][NumDots] = {{0}};
    int row, col;
    for (row = 0; row < BoardSize; row++) {
        for (col = 0; col < BoardSize; col++) {
			if (CONTAINS(mask, row, col)) {
                int start = INDEX(row, col);
                if (CountNeighbors(mask, row, col) < 2) {
                    buildPaths(mask, paths, seen, start, row, col, 0);
                } else {
                    Push(paths, (void*)DOTMASK(row, col));
                    seen[start][start] = 1;
                }
            }
        }
    }
}

void buildPaths(Mask mask, Queue *paths, int seen[NumDots][NumDots], int startIndex, int row, int col, Mask path) {
	mask = REMOVE(mask, row, col);
    path = ADD(path, row, col);

    int currentIndex = INDEX(row, col);
	if (!seen[startIndex][currentIndex]) {
		seen[startIndex][currentIndex] = 1;
		seen[currentIndex][startIndex] = 1;
        Push(paths, (void*)path);
	}

    if (CONTAINS(mask, row-1, col)) {
        buildPaths(mask, paths, seen, startIndex, row-1, col, path);
    }
    if (CONTAINS(mask, row+1, col)) {
        buildPaths(mask, paths, seen, startIndex, row+1, col, path);
    }
    if (CONTAINS(mask, row, col-1)) {
        buildPaths(mask, paths, seen, startIndex, row, col-1, path);
    }
    if (CONTAINS(mask, row, col+1)) {
        buildPaths(mask, paths, seen, startIndex, row, col+1, path);
    }
}

void PrintMask(Mask mask) {
    int row, col;
    for (row = 0; row < BoardSize; row++) {
        for (col = 0; col < BoardSize; col++) {
            if (CONTAINS(mask, row, col)) {
                printf(" *");
            } else {
                printf("  ");
            }
        }
        printf("\n");
    }
    printf("\n");
}
